### PR TITLE
Language Server: Basic Signature Help

### PR DESCRIPTION
**Pull request by ScottCarda-MS**
Adds basic signature help to the language server.

### CODE REVIEW

File: wasm/src/language_service.rs [hunk 1]
```diff
@@ -178,6 +208,47 @@ serializable_type! {
...
+
+serializable_type! {
+    ParameterInformation,
+    {
+        label: Span,
+        documentation: Option<String>,
+    },
+    r#"export interface IParameterInformation {
+        label: { start: number; end: number };
+        documentation: string | undefined;
```
Comment (minestarks) [inline wasm/src/language_service.rs:248]: You can use `ISpan` here which is now defined in this file. Also `?` for an optional parameter looks more consistent in style (though it's basically the same as `string | undefined`)

```suggestion
        label: ISpan;
        documentation?: string;
```

File: wasm/src/language_service.rs [hunk 2]
```diff
@@ -178,6 +208,47 @@ serializable_type! {
...
+serializable_type! {
+    SignatureInformation,
+    {
+        label: String,
+        documentation: Option<String>,
+        parameters: Vec<ParameterInformation>,
+    },
+    r#"export interface ISignatureInformation {
+        label: string;
+        documentation: string | undefined;
```
Comment (minestarks) [inline wasm/src/language_service.rs:235]: For consistency 

```suggestion
        documentation?: string;
```

File: wasm/src/language_service.rs [hunk 3]
```diff
@@ -178,6 +208,47 @@ serializable_type! {
     IDefinition
 }
 
+serializable_type! {
+    SignatureHelp,
+    {
+        signatures: Vec<SignatureInformation>,
+        active_signature: u32,
+        active_parameter: u32,
```
Comment (minestarks) [inline wasm/src/language_service.rs:216]: We should be able to use camelCase now without clippy yelling at us. (Another style consistency thing for this file)

File: vscode/src/extension.ts [hunk 4]
```diff
@@ -65,6 +66,16 @@ export async function activate(context: vscode.ExtensionContext) {
     )
   );
 
+  // signature help
+  context.subscriptions.push(
+    vscode.languages.registerSignatureHelpProvider(
+      "qsharp",
```
Comment (minestarks) [inline vscode/src/extension.ts:72]: Could you use `qsharpDocumentFilter` like we do for hover above? (And if you could fix go-to-def as well that would be cherry on top). 

(No practical difference today, but I have plans for this later when I work on notebook cells. Would be nice to have everything consistent now).

File: language_service/src/signature_help/tests.rs [hunk 5]
```diff
@@ -0,0 +1,623 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use super::get_signature_help;
+use crate::test_utils::{compile_with_fake_stdlib, get_source_and_marker_offsets};
+use expect_test::{expect, Expect};
+use indoc::indoc;
+
+/// Asserts that the hover text at the given cursor position matches the expected hover text.
```
Comment (minestarks) [inline language_service/src/signature_help/tests.rs:9]: Comment seems copy-pasted from hover, update the wording?

File: language_service/src/signature_help.rs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        match &decl.input.kind {
+            hir::PatKind::Discard | hir::PatKind::Bind(_) => {
+                vec![self.make_params_with_offset(offset, &decl.input)]
+            }
+            hir::PatKind::Tuple(items) => {
+                let mut cumulative_offset = offset;
+                items
+                    .iter()
+                    .map(|item| {
+                        let info = self.make_params_with_offset(cumulative_offset, item);
```
Comment (minestarks) [inline language_service/src/signature_help.rs:119]: Extreme nitpicking here, but I feel like this should be called `make_param_with_offset`, singular.

File: language_service/src/signature_help.rs [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+fn unwrap_parens(expr: &ast::Expr) -> &ast::Expr {
+    match &*expr.kind {
+        ast::ExprKind::Paren(inner) => unwrap_parens(inner),
+        _ => expr,
+    }
+}
+
+impl SignatureHelpFinder<'_> {
+    fn get_params(&self, decl: &hir::CallableDecl) -> Vec<ParameterInformation> {
+        let offset = self.display.get_param_offset(decl);
```
Comment (minestarks) [inline language_service/src/signature_help.rs:108]: Actually it'd be really helpful if you could add a comment on top that describes what we're doing with a code example. Such as:

```
// operation Foo(bar: Int, baz: Int)
//               ^param 1 start
//                         ^parameter 2 start
```

Feel free to modify my suggestion per your taste. I find that my eyes kind of glaze over all the Rust code but a visual example really gets across what the function is trying to accomplish.

**Review by minestarks — APPROVED** [review]
Looks great, left a bunch of style suggestions.

File: language_service/src/signature_help.rs (hunk 7 above)
Comment (ScottCarda-MS) [reply language_service/src/signature_help.rs:115]: I left a similar example as a comment along with a small description of what the function is doing.

File: language_service/src/signature_help/tests.rs (hunk 5 above)
Comment (ScottCarda-MS) [reply language_service/src/signature_help/tests.rs]: Oh good catch! Fixed

File: language_service/src/signature_help.rs (hunk 6 above)
Comment (ScottCarda-MS) [reply language_service/src/signature_help.rs]: done

File: vscode/src/extension.ts (hunk 4 above)
Comment (ScottCarda-MS) [reply vscode/src/extension.ts]: done

File: wasm/src/language_service.rs (hunk 3 above)
Comment (ScottCarda-MS) [reply wasm/src/language_service.rs]: Thank goodness!

File: wasm/src/language_service.rs (hunk 2 above)
Comment (ScottCarda-MS) [reply wasm/src/language_service.rs]: done

File: wasm/src/language_service.rs (hunk 1 above)
Comment (ScottCarda-MS) [reply wasm/src/language_service.rs]: done

**Review by swernli — APPROVED** [review]

**Pull request merged**